lib/improver/tests/nbhood/use_nbhood/test_CollapseMaskedNeighbourhoodCoordinate.py

```diff
@@ -0,0 +1,531 @@
+                                     [0.13, 0.17, 0.2, 0.2, 0.2],
+                                     [0.17, 0.19, 0.2, 0.2, 0.2],
+                                     [0.2, 0.2, 0.2, 0.19, 0.15],
+                                     [0.2, 0.2, 0.2, 0.16, 0.12]],
+                                    [[0.12, 0.13, 0.2, 0.2, 0.2],
+                                     [0.13, 0.17, 0.2, 0.2, 0.2],
+                                     [0.17, 0.19, 0.2, 0.2, 0.2],
+                                     [0.2, 0.2, 0.2, 0.19, 0.15],
+                                     [0.2, 0.2, 0.2, 0.16, 0.12]]])
+        result = self.plugin.process(nbhood_cube)
+        self.assertArrayAlmostEqual(expected_result, result.data)
+
+    def test_preserve_dimsensions_input(self):
+        """Test that the dimsensions on the output cube are the same as the
+           input cube, apart from the collapsed dimension.
+           Add threshold and realization coordinates and check they are in the
+           right place after collapsing the topographic_zone coordinate."""
+        nbhood_cube = self.weights_cube.copy()
+        nbhood_cube.remove_coord("realization")
+        nbhood_cubes = iris.cube.CubeList()
+        for i in range(3):
+            threshold_coord = DimCoord([i], long_name="threshold")
+            threshold_cube = iris.util.new_axis(nbhood_cube.copy())
+            threshold_cube.add_dim_coord(threshold_coord, 0)
+            nbhood_cubes.append(threshold_cube)
+        nbhood_cube = nbhood_cubes.concatenate_cube()
+        nbhood_cubes = iris.cube.CubeList()
+        for i in range(4):
+            realization_coord = DimCoord([i], long_name="realization")
+            realization_cube = iris.util.new_axis(nbhood_cube.copy())
+            realization_cube.add_dim_coord(realization_coord, 0)
+            nbhood_cubes.append(realization_cube)
+        nbhood_cube = nbhood_cubes.concatenate_cube()
+        result = self.plugin.process(nbhood_cube)
+        expected_dims = [coord for coord in nbhood_cube.dim_coords
+                         if coord.long_name is not "topographic_zone"]
+        self.assertEqual(result.dim_coords, tuple(expected_dims))
+        self.assertEqual(result.coord_dims("realization"), (0,))
+        self.assertEqual(result.coord_dims("threshold"), (1,))
+        self.assertEqual(result.coord_dims("projection_y_coordinate"), (2,))
+        self.assertEqual(result.coord_dims("projection_x_coordinate"), (3,))
+
+    def test_preserve_dimsensions_with_single_point(self):
+        """Test that the dimsensions on the output cube are the same as the
+           input cube, appart from the collapsed dimension.
+           Add threshold and realization coordinates and check they are in the
+           right place after collapsing the topographic_zone coordinate.
+           Check that a dimension coordinate with a single point is preserved
+           and not demoted to a scalar coordinate."""
```

Comment:
This and the previous test seem to have a lot of duplication, although I can see they are testing subtly different things, but I think they could be cleaned up.  The second test I think should not duplicate lines 479-483, but should just test the one thing - that the dimension coordinate is not demoted.  It would be clearer too if the duplication could be moved into a separate function (even setUp(), if appropriate).

This would probably fall out more naturally if process() were further broken down into some smaller functions (as suggested inline, largely because this test class has to be so large).  Then the required set up would be more consistent over the tests in each class.

Reply:
Again I've tidied up this a bit. Let me know if you think it goes far enough.